wire-library/wire-compiler/src/main/java/com/squareup/wire/Manifest.kt

```diff
@@ -0,0 +1,310 @@
+import com.squareup.wire.schema.Type
+import com.squareup.wire.schema.internal.DagChecker
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.builtins.MapSerializer
+import kotlinx.serialization.builtins.serializer
+import java.nio.file.Path
+import java.util.ArrayDeque
+
+data class Manifest(
+  val modules: Map<String, Module>
+) {
+  init {
+    val dagChecker = DagChecker(modules.keys) { moduleName ->
+      modules.getValue(moduleName).dependencies
+    }
+    val cycles = dagChecker.check()
+    require(cycles.isEmpty()) {
+      buildString {
+        append("ERROR: Manifest modules contain dependency cycles:\n")
+        for (cycle in cycles) {
+          append(" - ")
+          append(cycle)
+          append('\n')
+        }
+      }
+    }
+  }
+
+  /** Modules which have no dependencies. */
+  val roots: Set<String> = modules.filterValues { it.dependencies.isEmpty() }.keys
+
+  /** Module ordering such that each modules dependencies appear prior to itself. */
+  val order: List<String> = run {
+    val seen = LinkedHashSet<String>() // Insertion order is important to produce the final list!
+    val queue = ArrayDeque<String>().apply { addAll(roots) }
+    while (queue.isNotEmpty()) {
+      val currentName = queue.removeFirst()
+      val currentDependencies = modules.getValue(currentName).dependencies
+      if (seen.containsAll(currentDependencies)) {
+        seen += currentName
+
+        val downstreamModules = modules.filterValues { currentName in it.dependencies }.keys
+        queue += downstreamModules
+      } else {
+        // All dependencies have not been seen so move to the back of the line to try again later.
+        queue += currentName
+      }
+    }
+    seen.toList()
+  }
+
+  @Serializable
+  data class Module(
+    val dependencies: Set<String> = emptySet(),
+    val roots: Set<String> = setOf("*"),
+    val prunes: Set<String> = emptySet()
+  )
+
+  companion object {
+    @JvmStatic
+    val NONE = Manifest(
+        modules = mapOf("./" to Module())
+    )
+
+    private val serializer = MapSerializer(String.serializer(), Module.serializer())
+
+    fun fromYaml(string: String): Manifest {
+      return Manifest(Yaml.default.parse(serializer, string))
+    }
+  }
+}
+
+/** Module name to partition info. The iteration order of this map is the generation order. */
+class PartitionedSchema(
+  modules: Map<String, Partition>,
+  val warnings: List<String>,
+  val errors: List<String>
+) : Map<String, Partition> by modules {
+  class Partition(
+    val schema: Schema,
+    val types: Set<ProtoType>,
+    val upstreamTypes: Map<ProtoType, String>
+  )
+}
+
+private const val debug = false
+
+fun Schema.partition(manifest: Manifest): PartitionedSchema {
+  val modules = manifest.modules
+
+  if (debug) {
+    println("Modules: ${modules.keys}")
+    println("Roots: ${manifest.roots}")
+    println("Order: ${manifest.order}")
+  }
+
+  val disjointGraphs = manifest.roots
```

Comment:
consider doing like DagChecker and splitting the graph theory work from the application of that work

Reply:
(a bug in disjoinGraphs computation is pretty brutal to debug the way this function is structured)

Reply:
Split into its own type!